src/mysky/index.ts

```diff
@@ -0,0 +1,302 @@
+        if (uiConnection) {
+          uiConnection.close();
+        }
+
+        // Clean up the event listeners and promises.
+        controllerError.cleanup();
+      });
+
+    // Return.
+    if (this.pendingPermissions.length > 0) {
+      return false;
+    }
+    return true;
+  }
+
+  async userID(): Promise<string> {
+    return this.connector.connection.remoteHandle().call("userID");
+  }
+
+  async getJSON(path: string, opts?: CustomGetJSONOptions): Promise<JsonData | null> {
+    // TODO: Check for valid inputs.
+
+    const publicKey = await this.userID();
+    const dataKey = deriveDiscoverableTweak(path);
+
+    return this.connector.client.db.getJSON(publicKey, Buffer.from(dataKey).toString(), opts);
+  }
+
+  async setJSON(path: string, json: JsonData, revision?: bigint, opts?: CustomSetJSONOptions): Promise<void> {
+    // TODO: Check for valid inputs.
+
+    const publicKey = await this.userID();
+    const dataKey = deriveDiscoverableTweak(path);
+
+    const entry = await getOrCreateRegistryEntry(
+      this.connector.client,
+      hexToUint8Array(publicKey),
+      Buffer.from(dataKey).toString(),
+      json,
+      revision,
+      opts
+    );
+
+    const signature = await this.signRegistryEntry(entry, path);
+
+    return await this.connector.client.registry.postSignedEntry(hexToUint8Array(publicKey), entry, signature, opts);
+  }
+
+  // ================
+  // Internal Methods
+  // ================
+
+  protected async catchUiError(errorMsg: string) {
+    this.uiError = errorMsg;
+  }
+
+  protected async launchUi(): Promise<[Window, Connection]> {
+    const options = this.connector.options;
+    const mySkyUrl = this.connector.url;
+    const uiUrl = `${mySkyUrl}/${mySkyUiRelativeUrl}`;
+
+    // Open the window.
+
+    const childWindow = window.open(uiUrl);
+    if (!childWindow) {
+      throw new Error(`Could not open window at '${uiUrl}'`);
+    }
+
+    // Complete handshake with UI window.
+
+    const messenger = new WindowMessenger({
+      localWindow: window,
+      remoteWindow: childWindow,
+      remoteOrigin: "*",
+    });
+    const methods = {
+      catchUiError: this.catchUiError,
+    };
+    // TODO: Get handshake values from optional fields.
+    const connection = await ParentHandshake(
+      messenger,
+      methods,
+      options.handshakeMaxAttempts,
+      options.handshakeAttemptsInterval
+    );
+
+    return [childWindow, connection];
+  }
+
+  protected async loadDac(dac: DacLibrary): Promise<void> {
+    // Initialize DAC.
+    await dac.init(this.connector.client, this.connector.options);
+
+    // Add DAC permissions.
+    const perms = await dac.getPermissions();
+    this.addPermissions(...perms);
+  }
+
+  // TODO: Move to promise.ts file in skynet-mysky-utils?
+  /**
+   * Checks if there has been an error from the UI on an interval.
+   */
+  protected monitorUiError(): { promise: Promise<void>; controller: PromiseController } {
+    const pingInterval = 100;
```

Comment:
if not defined as a constant up top, might as well pass it in as a parameter that defaults to `100`

Reply:
Maybe later.

Reply:
https://github.com/SkynetHQ/skynet-js/issues/28